gravitee-apim-console-webui/src/management/application/details/user-group-access/groups/application-general-groups.component.ts

```diff
@@ -32,48 +32,149 @@ import { SnackBarService } from '../../../../../services-ngx/snack-bar.service';
   styleUrls: ['./application-general-groups.component.scss'],
 })
 export class ApplicationGeneralGroupsComponent implements OnInit, OnDestroy {
-  public groups: Group[];
+  public groups: Group[] = [];
   public application: Application;
   public form: FormGroup;
   public initialFormValue: unknown;
   public isReadonly = false;
 
+  public page = 1;
+  public pageSize = 50;
+  public hasMoreGroups = true;
+  public isLoading = false;
+
   private unsubscribe$: Subject<void> = new Subject<void>();
+  private scrollContainer: HTMLElement | null = null;
+  private scrollListener: (() => void) | null = null;
+  private totalPages = 0;
 
   constructor(
-    private readonly groupService: GroupService,
+    private readonly groupv2Service: GroupV2Service,
     private readonly applicationService: ApplicationService,
     private readonly activatedRoute: ActivatedRoute,
     private readonly snackBarService: SnackBarService,
   ) {}
 
   ngOnInit() {
-    combineLatest([this.applicationService.getById(this.activatedRoute.snapshot.params.applicationId), this.groupService.list()])
+    this.applicationService
+      .getById(this.activatedRoute.snapshot.params.applicationId)
       .pipe(
-        tap(([application, groups]) => {
-          this.groups = groups;
+        tap((application) => {
           this.application = application;
           this.isReadonly = application.origin === 'KUBERNETES';
         }),
         takeUntil(this.unsubscribe$),
       )
       .subscribe(() => {
-        const userGroupList: Group[] = this.groups.filter((group) => this.application.groups?.includes(group.id));
+        this.loadInitialGroups();
+      });
+  }
+
+  private updatePagination(pagination: any): void {
+    if (!pagination) return;
+    this.totalPages = pagination.pageCount;
+    if (pagination.page >= this.totalPages) {
+      this.hasMoreGroups = false;
+    } else {
+      this.page = pagination.page + 1;
+    }
+  }
+
+  private mergeGroups(newGroups: Group[]): void {
+    const selectedIds = new Set(this.application.groups ?? []);
+    const existingIds = new Set(this.groups.map((g) => g.id));
+    const filteredNewGroups = newGroups.filter((g) => !existingIds.has(g.id));
+
+    this.groups = [
+      ...this.groups.filter((g) => selectedIds.has(g.id)),
+      ...this.groups.filter((g) => !selectedIds.has(g.id)),
+      ...filteredNewGroups,
+    ];
```

Comment:
Can this be simplified to this?
```suggestion
    this.groups = [
      ...this.groups,
      ...filteredNewGroups,
    ];
```

Reply:
I'd prefer not to simply this because:

That line does two important things:

1. Reorders the existing groups so that selected groups appear first in the list.
2. Appends any newly fetched groups (filteredNewGroups) after the existing ones, without duplicates.

Reply:
When mergeGroups() is called, we’re overwriting this.groups each time.
If any step in our lifecycle (e.g. getSelectedGroupData()) calls it twice — for instance:

this.groups = [...selectedGroups];
this.mergeGroups(page1Groups);

Then:
->this.groups already holds selected groups.
->filteredNewGroups is calculated relative to that partial list.
->->->But if page1Groups overlaps with selected groups (which it often does), some duplicates can slip in because:
->selectedGroups were set before filtering, and page1Groups include the same items again.

So filteredNewGroups doesn’t actually filter them out (since existingIds was based only on selectedGroups, not full groups).

<img width="1693" height="811" alt="image" src="https://github.com/user-attachments/assets/16f85c58-a18c-489b-912b-3e4dec403dc7" />

Reply:
you may try optimize this .filter 
It will iterate only once and do the same IMHO. 
```// Create temporary arrays for partitioned groups
const selectedGroups = [];
const unselectedGroups = [];

// Partition the existing groups in a single pass
for (const group of this.groups) {
  if (selectedIds.has(group.id)) {
    selectedGroups.push(group);
  } else {
    unselectedGroups.push(group);
  }
}

// Reconstruct the array
this.groups = [
  ...selectedGroups,
  ...unselectedGroups,
  ...filteredNewGroups,
];```

Reply:
Yes, it is better in performance. I have pushed the suggested code changes.